File: src/com/damnae/osukeysoundsplitter/pathprovider/AdditionsKeysoundPathProvider.java
package com.damnae.osukeysoundsplitter.pathprovider;

public class AdditionsKeysoundPathProvider extends BaseKeysoundPathProvider {
	private int sampleType;
	private int samplesetIndex;
	private int additionIndex;

	public AdditionsKeysoundPathProvider(int initialSampleType) {
		sampleType = Math.max(1, initialSampleType);
	}

	@Override
	protected String getNewKeysoundPath(String extension) {
		String path = buildPath(extension);
		incrementPath();
		return path;
	}

	private String buildPath(String extension) {
		return samplesetNames[samplesetIndex] + "-hit"
				+ additionNames[additionIndex]
				+ (sampleType > 1 ? sampleType : "") + "." + extension;
	}

	private void incrementPath() {
		++additionIndex;
		if (additionIndex >= additionNames.length) {
			additionIndex = 0;
			++samplesetIndex;
			if (samplesetIndex >= samplesetNames.length) {
				samplesetIndex = 0;
				++sampleType;
			}
		}
	}

	public static int getSampleType(String path) {
		int endPosition = path.lastIndexOf('.');
		for (int i = endPosition - 1; i >= 0; --i) {
			char c = path.charAt(i);
			if (!(c >= '0' && c <= '9'))
				return Integer.valueOf(path.substring(i + 1, endPosition));

		}
		return 0;
	}

	public static int getAdditionsSampleset(String path) {
		for (int i = 0, size = samplesetNames.length; i < size; ++i) {
			if (path.startsWith(samplesetNames[i]))
				return i + 1;
		}
		return 0;
	}

	public static int getAdditions(String path) {
		for (int i = 0, size = additionNames.length; i < size; ++i) {
			if (path.contains(additionNames[i]))
				return additionCodes[i];
		}
		return 0;
	}

	private static String[] samplesetNames = { "normal", "soft", "drum" };
	private static String[] additionNames = { "whistle", "finish", "clap" };
	private static int[] additionCodes = { 2, 4, 8 };
}

File: osu!KeysoundSplitter/PathProvider/AdditionsKeysoundPathProvider.cs
using System;

namespace osuKeysoundSplitter.PathProvider
{
    public class AdditionsKeysoundPathProvider : BaseKeysoundPathProvider
    {
        private int sampleSet;
        private int sampleTypeIndex;
        private int additionIndex;

        public AdditionsKeysoundPathProvider(KeysoundCache keysoundCache,
                int initialSampleType)
            : base(keysoundCache)
        {
            sampleSet = Math.Max(1, initialSampleType);
        }

        protected override string getNewKeysoundPath(string extension)
        {
            string path = buildPath(extension);
            incrementPath();
            return path;
        }

        private String buildPath(String extension)
        {
            return sampleTypeNames[sampleTypeIndex] + "-hit"
                    + additionNames[additionIndex]
                    + (sampleSet > 1 ? sampleSet.ToString() : "") + "." + extension;
        }

        private void incrementPath()
        {
            ++additionIndex;
            if (additionIndex >= additionNames.Length)
            {
                additionIndex = 0;
                ++sampleTypeIndex;
                if (sampleTypeIndex >= sampleTypeNames.Length)
                {
                    sampleTypeIndex = 0;
                    ++sampleSet;
                }
            }
        }

        protected override bool isPathValid(String keysoundPath)
        {
            return !keysoundPath.Contains("/") && !keysoundPath.Contains("\\")
                    && keysoundPath.Contains("-hit")
                    && getSampleType(keysoundPath) != 0
                    && getAdditions(keysoundPath) != 0;
        }

        public static int getSampleSet(String path)
        {
            int endPosition = path.LastIndexOf('.');
            for (int i = endPosition - 1; i >= 0; --i)
            {
                char c = path[i];
                if (!(c >= '0' && c <= '9'))
                    return int.Parse(path.Substring(i + 1, endPosition));
            }
            return 0;
        }

        public static int getSampleType(String path)
        {
            for (int i = 0, size = sampleTypeNames.Length; i < size; ++i)
            {
                if (path.StartsWith(sampleTypeNames[i]))
                    return i + 1;
            }
            return 0;
        }

        public static int getAdditions(String path)
        {
            for (int i = 0, size = additionNames.Length; i < size; ++i)
            {
                if (path.Contains(additionNames[i]))
                    return additionCodes[i];
            }
            return 0;
        }

        private static String[] sampleTypeNames = { "normal", "soft", "drum" };
        private static String[] additionNames = { "whistle", "finish", "clap" };
        private static int[] additionCodes = { 2, 4, 8 };
    }
}